src/background/telemetry/exception-telemetry-listener.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { BaseActionPayload } from 'background/actions/action-payloads';
+import { TelemetryEventHandler } from 'background/telemetry/telemetry-event-handler';
+import { Logger } from 'common/logging/logger';
+import * as TelemetryEvents from '../../common/extension-telemetry-events';
+import { UnhandledExceptionTelemetryData } from '../../common/extension-telemetry-events';
+
+export class ExceptionTelemetryListener {
+    constructor(private readonly telemetryEventHandler: TelemetryEventHandler) {}
+
+    public initialize(
+        logger: Logger,
+        extWindow: Window = window,
+        extConsole: Console = console,
+    ): void {
+        const sendExceptionTelemetry = (message: string, stackTrace?: string): void => {
+            const telemetry: UnhandledExceptionTelemetryData = { message, stackTrace };
+            const payload: BaseActionPayload = {
+                telemetry,
+            };
+
+            this.telemetryEventHandler.publishTelemetry(
+                TelemetryEvents.UNHANDLED_EXCEPTION,
+                payload,
+            );
+        };
+
+        // Catch top level synchronous errors
+        extWindow.onerror = function (
```

Comment:
I think all of these hooks should be taking care to avoid reentrancy (ie, to avoid infinite recursion in the event that `sendExceptionTelemetry` somehow ends up logging an error/throwing an unhandled exception/floating a rejecting promise). The browser makes *some* attempts at avoiding infinite recursion here (eg, neither Chromium nor Firefox will trigger an `onerror` handler a second time if the `onerror` handler itself throws an error), but it isn't absolute and it isn't mandated by the standard, so we shouldn't rely on it (eg, Firefox doesn't protect against `onerror` and `onunhandledrejection` bouncing back and forth forever).

I would recommend protecting against this by having the listener maintain a `hookIsActive` boolean and having each individual hook be of the form:

```ts
if (this.hookIsActive) { return; }
this.hookIsActive = true;
try {
    // insert current hook implementation
} finally {
    this.hookIsActive = false;
}
```